server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/DistributionPlanner.java

```diff
@@ -29,7 +39,214 @@ public DistributionPlanner(Analysis analysis, LogicalQueryPlan logicalPlan) {
+      return null;
+    }
+
+    public PlanNode visitTimeJoin(TimeJoinNode node, DistributionPlanContext context) {
+      TimeJoinNode root = (TimeJoinNode) node.clone();
+
+      // Step 1: Get all source nodes. For the node which is not source, add it as the child of
+      // current TimeJoinNode
+      List<SeriesScanNode> sources = new ArrayList<>();
+      for (PlanNode child : node.getChildren()) {
+        if (child instanceof SeriesScanNode) {
+          // If the child is SeriesScanNode, we need to check whether this node should be seperated
+          // into several splits.
+          SeriesScanNode handle = (SeriesScanNode) child;
+          Set<DataRegion> dataDistribution =
+              analysis.getPartitionInfo(handle.getSeriesPath(), handle.getTimeFilter());
+          // If the size of dataDistribution is m, this SeriesScanNode should be seperated into m
+          // SeriesScanNode.
+          for (DataRegion dataRegion : dataDistribution) {
+            SeriesScanNode split = (SeriesScanNode) handle.clone();
+            split.setDataRegion(dataRegion);
+            sources.add(split);
+          }
+        } else if (child instanceof SeriesAggregateScanNode) {
+          // TODO: (xingtanzjr) We should do the same thing for SeriesAggregateScanNode. Consider to
+          // make SeriesAggregateScanNode
+          // and SeriesScanNode to derived from the same parent Class because they have similar
+          // process logic in many scenarios
+        } else {
+          // In a general logical query plan, the children of TimeJoinNode should only be
+          // SeriesScanNode or SeriesAggregateScanNode
+          // So this branch should not be touched.
+          root.addChild(visit(child, context));
+        }
+      }
+
+      // Step 2: For the source nodes, group them by the DataRegion.
+      Map<DataRegion, List<SeriesScanNode>> sourceGroup = new HashMap<>();
+      sources.forEach(
+          source -> {
+            List<SeriesScanNode> group =
+                sourceGroup.containsKey(source.getDataRegion())
+                    ? sourceGroup.get(source.getDataRegion())
+                    : new ArrayList<>();
+            group.add(source);
+            sourceGroup.put(source.getDataRegion(), group);
+          });
+
+      // Step 3: For the source nodes which belong to same data region, add a TimeJoinNode for them
+      // and make the
+      // new TimeJoinNode as the child of current TimeJoinNode
+      sourceGroup.forEach(
+          (dataRegion, seriesScanNodes) -> {
+            if (seriesScanNodes.size() == 1) {
+              root.addChild(seriesScanNodes.get(0));
+            } else {
+              // We clone a TimeJoinNode from root to make the params to be consistent
+              TimeJoinNode parentOfGroup = (TimeJoinNode) root.clone();
+              seriesScanNodes.forEach(parentOfGroup::addChild);
+              root.addChild(parentOfGroup);
+            }
+          });
```

Comment:
What if all `seriesScanNodes.size()` > 1, we will have a duplicated TimeJoinNode. We may simply choose the first entry of `sourceGroup` to generate the root `TimeJoinNode` and then add others as its children.

Reply:
Yes, we can do some optimization here

Reply:
For the scenario you mentioned, I think we should put the `duplicated` TimeJoinNode in current stage, although we should delete it if the `duplicated` one and his parent are in the same Fragment. we can make the optimization in following step.

By the way, your comment reminds me another optimization here. That is, if we have only one element in sourceGroup, we will produce a duplicated TimeJoinNode...

Reply:
I added a TODO here to remind us the optimization. Because this code is not really completed and let's complete it in later PR